### PR TITLE
partial call chain and on-the-fly verification

### PR DESCRIPTION
**Pull request by lalilalalalu**

### CODE REVIEW

File: tools/verifyio/verifyio.py [hunk 1]
```diff
@@ -285,22 +377,25 @@ def build_call_chain_str(call_chain, reader):
     #print('6. RAM Used (GB):', psutil.virtual_memory()[3]/1000000000)
     print("Step 2. match mpi calls: %.3f secs, mpi edges: %d" %((t2-t1),len(mpi_edges)))
 
-    t1 = time.time()
-    vio.G = VerifyIOGraph(all_nodes, mpi_edges, include_vc=True)
-    t2 = time.time()
-    print("Step 3. build happens-before graph: %.3f secs, nodes: %d" %((t2-t1), vio.G.num_nodes()))
-    #print('7. RAM Used (GB):', psutil.virtual_memory()[3]/1000000000)
+    if vio.without_graph:
```
Comment (wangvsa) [inline tools/verifyio/verifyio.py:380]: we don't need this extra argument. you could simply check if vio.algorithm == 4.
only that algorithm does not need the graph.

Comment (wangvsa) [inline tools/verifyio/verifyio.py]: also, are you swapped the if condition? you are building the graph under if vio.without_graph.

File: tools/verifyio/verifyio.py [hunk 2]
```diff
@@ -252,6 +344,8 @@ def build_call_chain_str(call_chain, reader):
                         default=3, help="1: graph reachibility, 2: transitive closure, 3: vector clock, 4: on-the-fly MPI check")
     parser.add_argument("--show_details", action="store_true", help="Show details of the conflicts")
     parser.add_argument("--show_summary", action="store_true", help="Show summary of the conflicts")
+    parser.add_argument("--show_full_chain", action="store_true", help="Show the full call chain of the conflicts")
+    parser.add_argument("--without_graph", action="store_false", help="Use graph to verify the semantics")
```
Comment (wangvsa) [inline tools/verifyio/verifyio.py]: don't need this argument. you could check the algorithm. algorithm == 4 means no graph.

File: tools/verifyio/verifyio.py [hunk 3]
```diff
@@ -90,6 +114,17 @@ def verify_pair_proper_synchronization(n1, n2, vio):
 
     # Algorithm 4: On-the-fly MPI check
     if vio.algorithm == 4:
```
Comment (wangvsa) [inline tools/verifyio/verifyio.py:106]: for commit/posix semantics, next_sync_index is null.

File: tools/verifyio/verifyio.py [hunk 4]
```diff
@@ -62,11 +80,17 @@ def verify_pair_proper_synchronization(n1, n2, vio):
         v1 = vio.G.next_hb_node(n1, ["fsync", "close", "fclose"], rank)
         v2 = n2
     elif vio.semantics == "Session":
-        v1 = vio.G.next_po_node(n1, ["close", "fclose", "fsync"])
-        v2 = vio.G.prev_po_node(n2, ["open",  "fopen",  "fsync"])
+        if vio.G is None:
+            v1, v2, next_sync_index = find_next_prev_sync(n1, n2, all_nodes, ["close", "fclose", "fsync"], ["open",  "fopen",  "fsync"])
```
Comment (wangvsa) [inline tools/verifyio/verifyio.py]: no need to do this. once you have v1 and v2 (from next_po_node/prev_po_node), you could get the index using `index = self.G.nodes[v1.graph_key()]['index']`

**Review by wangvsa — CHANGES_REQUESTED** [review]

**Review by wangvsa — CHANGES_REQUESTED** [review]
plsase fix these.

File: tools/verifyio/verifyio.py (hunk 2 above)
Comment (lalilalalalu) [reply tools/verifyio/verifyio.py]: is done

File: tools/verifyio/verifyio.py (hunk 3 above)
Comment (lalilalalalu) [reply tools/verifyio/verifyio.py:106]: fixed

**Pull request merged**